include/dslash.h

```diff
@@ -77,9 +78,31 @@ namespace quda
       strncat(aux[kernel_type], aux_base, TuneKey::aux_n - 1);
     }
 
-    virtual bool tuneGridDim() const { return false; }
+    virtual bool tuneGridDim() const { return arg.kernel_type == EXTERIOR_KERNEL_ALL && arg.shmem > 0; }
     virtual unsigned int minThreads() const { return arg.threads; }
 
+    virtual unsigned int minGridSize() const
+    {
+      if (arg.kernel_type == EXTERIOR_KERNEL_ALL && arg.shmem > 0) {
+        int nDimComms = 0;
+        for (int d = 0; d < in.Ndim(); d++) nDimComms += arg.commDim[d];
+        return ((deviceProp.multiProcessorCount) / (2 * nDimComms)) * (2 * nDimComms);
+      } else {
+        return TunableVectorYZ::minGridSize();
+      }
+    }
+
+    virtual int gridStep() const
```

Comment:
Might be nice to add some comments here to document how the grid size is tuned, and what the constraints are when nvshmem is enabled.

Reply:
added a few more comments on this.